### DISCUSSION TITLE
A bit of seasonal fun: how well do you know Python?

### DISCUSSION THREAD

**Discussion by peterhinch** [micropython/micropython#13256] [Ideas]
Can you **predict** what this will do? Will it throw, and if so on which line? Otherwise, what output will it produce?
```python
class Foo:
    bar = 0
    def __init__(self):
        self.bar += 1

foo_instance = Foo()
print(Foo.bar)
print(foo_instance.bar)
```
This is not a MicroPython bug. Other simple Python gotchas welcome.

**Comment by Josverl** [comment]
I think it will throw a HollidayException("The self.bar is not open yet")

**Comment by GitHubsSilverBullet** [comment]
> This is not a MicroPython bug. Other simple Python gotchas welcome.

Interesting.  I would have said that it throws the moment the instance is created.
Apparently as there's no instance variable (self.bar) to increment, __init__() seems to use the class variable instead.
That's nasty. ;-)

**Reply by rkompass** [reply]
I had the same thought and was surprised as well.
Thank you Peter for an interesting insight.
Of course the final Foo.bar value is now ???
`print(Foo.bar)`

**Reply by GitHubsSilverBullet** [reply]
Looks like
```python
        self.bar += 1
        self.bar = Foo.bar + 1
        self.bar = __class__.bar + 1
```
are all the same. Though for clarity, I prefer the second one

**Comment by sosi-deadeye** [comment]
The code assigns `1` to the new instance variable `self.bar`, instead to the class variable `bar`.
If you do not assign `self.bar`, you've read access to the class variable through the instance.

Here an example:
```python
import time


class Foo:
    _instances = set()

    @property
    def instances(self) -> int:
        """
        Return the number of instances
        """
        return len(self._instances)

    def __init__(self, name):
        # the static method __new__ could be
        # used to modify the class before returning it

        self._instances.add(self)
        # alternative, if assignment is required
        # cls = self.__class__
        # cls._instances = 42

        print(f"Foo has now {self.instances} instances")
        self.name = name
        self.timestamp = time.time()

    # __del__ is not called in Micropython,
    # but this works with CPython
    def remove(self):
        self._instances.remove(self)
        print(f"Foo has now {self.instances} instances")
        print(
            f"Removed {self.name} from instances of Foo after {time.time() - self.timestamp:.2f} s"
        )


print()
print("Instances of Foo:", Foo._instances)
f1 = Foo("First")
print("Instances of Foo:", Foo._instances)
print()
f2 = Foo("Second")
print("Instances of Foo:", Foo._instances)
print()
time.sleep(1)
f1.remove()
print("Instances of Foo:", Foo._instances)
print()
time.sleep(1)
f2.remove()
print("Instances of Foo:", Foo._instances)
```

**Reply by sosi-deadeye** [reply]
Additionally: This Code works because `_instances` is _mutated_ and not _assigned_.

**Reply by peterhinch** [reply]
I think the moral of the story is always to reference a class variable by classname, and not by `self`:
```py
class Foo:
    bar = 0
    def __init__(self):
       Foo.bar += 1
```
Season's greetings to all!

**Comment by rkompass** [comment]
O.K., another one:
```
nlist = [[]] * 4
nlist[0].append('hello')
nlist[1].append('world!')
print(nlist)
```
will print what?

**Reply by peterhinch** [reply]
Oh yes, I've encountered this before. Clue: what will this print?
```python
>>> nlist[0] is nlist[1]
```
It's a very odd Python feature.

**Comment by shariltumin** [comment]
```py
class K: pass

def a(o):
    o.n **= 2
    return o.n
def b(o,n):
    o.n = int(n**(0.5))
    return o.n

K.n = 2023; K.a = a; K.b = b

k0 = K()
print(k0.n == k0.b(k0.a()))

K.n = 2024

k1 = K()
print(k1.b(k1.a()) == k1.n)

print(k0.n == k1.n)
print(k0.n, '-->', k1.n)
```

Merry Christmas and Happy New Year.

**Reply by peterhinch** [reply]
I think this is another manifestation of the behaviour highlighted in the OP.
